src/main.rs

```diff
@@ -491,6 +590,41 @@ impl From<CheckRelease> for cargo_semver_checks::Check {
     }
 }
 
+/// Helper function to encapsulate the logic of validating that unstable options
+/// were not used without `-Z unstable-options` and issuing deprecation warnings
+/// for any stable feature flags that were explicitly specified.
+fn validate_feature_flags(config: &mut GlobalConfig, unstable_options: &mut UnstableOptions) {
+    // needed to avoid borrow checker errors when printing with config.
+    let stable_flags: Vec<_> = config
+        .feature_flags()
+        .iter()
+        .filter(|x| x.stable)
+        .copied()
+        .collect();
+
+    for stable_flag in stable_flags {
+        config
+            .shell_warn(format_args!(
+                "the feature flag {} has been stabilized and may be removed
+            from the list of feature flags in a future release.",
+                stable_flag.id
+            ))
+            .expect("printing failed");
+    }
+
+    if !config.feature_flag_enabled(FeatureFlag::UNSTABLE_OPTIONS)
+        && unstable_options != &UnstableOptions::default()
+    {
+        config
+            .shell_warn(
+                "unstable options were passed without `-Z unstable-options`. They will be ignored.",
+            )
+            .expect("print failed");
+
+        *unstable_options = UnstableOptions::default();
```

Comment:
Do `cargo` and `rustdoc` simply ignore unstable options, or do they exit non-zero? It feels like the latter might be safer, especially for a tool that's intended to be used in a CI environment where ~nobody reads the logs...

Reply:
Oh, you're right, at least `cargo` exits with an error instead.  Let's use that behavior.